Material chapter (#121)
festim-dev/FESTIM-workshop: book/content/material/material_advanced.md

---
jupytext:
  formats: ipynb,md:myst
  text_representation:
    extension: .md
    format_name: myst
    format_version: 0.13
    jupytext_version: 1.17.3
kernelspec:
  display_name: festim-workshop
  language: python
  name: python3
---

> # Advanced Material Features #

Author: ```suggestion
# Advanced functionality #
```

---
jupytext:
  formats: ipynb,md:myst
  text_representation:
    extension: .md
    format_name: myst
    format_version: 0.13
    jupytext_version: 1.17.3
kernelspec:
  display_name: festim-workshop
  language: python
  name: python3
---

# Advanced Material Features #

+++

This tutorial introduces more advanced ways of handling materials in FESTIM. A key addition in FESTIM 2.0 is the ability to perform multi-species simulations, which often requires assigning distinct material properties to each species (see __[documentation](https://festim.readthedocs.io/en/v2.0-alpha.3/api/index.html#festim.material.Material)__ for more information). In some cases, users may also need to define custom material properties—for example, specifying a turbulence-dependent viscosity or user-defined diffusivity.

Objectives:
* Assign different material properties to different species
* Solve a problem with custom material properties (e.g., diffusivity)

+++

> ## Assigning different material properties in a multi-species simulation ##

Author: ```suggestion
## Defining species-dependent material properties
```

> Consider the following 1D example that simulates the diffusion of protium, deuterium and tritium species:

Author: I think that if you write "Consider this" then that means you think this is important. In that case, it shouldn't be hidden. imo

> Consider the following 1D example that simulates the diffusion of protium, deuterium and tritium species:

Author: ```suggestion
Consider the following 1D example that simulates the diffusion of protium, deuterium and tritium:
```

```{code-cell} ipython3
:tags: [hide-input]

import festim as F
import numpy as np

my_model = F.HydrogenTransportProblem()

protium = F.Species("H")
deuterium = F.Species("D")
tritium = F.Species("T")
my_model.species = [protium, deuterium, tritium]

my_model.mesh = F.Mesh1D(np.linspace(0, 1, 100))

left_surf = F.SurfaceSubdomain1D(id=1, x=0)
right_surf = F.SurfaceSubdomain1D(id=2, x=1)
```

> Here we define the different diffusivities for each species. The `mat` dictionary stores keys for each species' diffusion coefficient `D_0` and activation energy `E_D`. We define one domain for all species called `bulk` that serves as the volume for the material:

Author: ```suggestion
Here we define the different diffusivities (or rather the diffusivities pre-exponential factors) for each species.
Instead of passing a single value to the `D_0` of `Material`, we pass a dictionary where the keys are the different `Species` objects and values are values of pre-exponential factors `D_0` and activation energy `E_D`.
We define one domain called `bulk` and attach the material to it:
```

```{code-cell} ipython3
mat = F.Material(

>     D_0={protium: 1.0e-3, deuterium: 3.0e-3, tritium: 5.0e-3},  # example diffusivity values for H, D, T

Author: No need for comments if you already explain it in the markdown cells

```suggestion
    D_0={protium: 1.0e-3, deuterium: 3.0e-3, tritium: 5.0e-3},
```

    E_D={protium: 0.0, deuterium: 0.0, tritium: 0.0},
)

> # One bulk domain using this material

Author: ```suggestion
```

bulk = F.VolumeSubdomain1D(id=1, borders=[0, 1], material=mat)

my_model.subdomains = [bulk, left_surf, right_surf]
```

To illustrate how different species' diffusion properties affect the simulation, we set the same boundary condition for each species (1 on the left boundary, 0 on the right). Now we can run the simulation and look at the concentration profile for each species:

```{code-cell} ipython3

> :tags: [hide-input]

Author: Same comment as above, if you explain this in the markdown cell, then don't hide it.

I think here you want to split this cell after `my_model.run()` and then only hide the post-processing cell

# Boundary conditions
my_model.boundary_conditions = [
    F.FixedConcentrationBC(left_surf, value=1, species=protium),
    F.FixedConcentrationBC(right_surf, value=0, species=protium),
    F.FixedConcentrationBC(left_surf, value=1, species=deuterium),
    F.FixedConcentrationBC(right_surf, value=0, species=deuterium),
    F.FixedConcentrationBC(left_surf, value=1, species=tritium),
    F.FixedConcentrationBC(right_surf, value=0, species=tritium),
]

my_model.temperature = 300
my_model.settings = F.Settings(atol=1e-10, rtol=1e-10, final_time=5)

my_model.settings.stepsize = F.Stepsize(1)
my_model.initialise()
my_model.run()

import matplotlib.pyplot as plt

def plot_profile(species, **kwargs):
    c = species.post_processing_solution.x.array[:]
    x = species.post_processing_solution.function_space.mesh.geometry.x[:,0]
    return plt.plot(x, c, **kwargs)

for species in my_model.species:
    plot_profile(species, label=species.name)

plt.xlabel('Position')
plt.ylabel('Concentration')
plt.legend()
plt.show()
```

Compare this to the case where each species has the same diffusion properties:

```{code-cell} ipython3
:tags: [hide-input]

> import festim as F
> import numpy as np
> 
> my_model = F.HydrogenTransportProblem()
> 
> protium = F.Species("H")
> deuterium = F.Species("D")
> tritium = F.Species("T")
> my_model.species = [protium, deuterium, tritium]
> 
> my_model.mesh = F.Mesh1D(np.linspace(0, 1, 100))
> 
> left_surf = F.SurfaceSubdomain1D(id=1, x=0)
> right_surf = F.SurfaceSubdomain1D(id=2, x=1)
> 
> mat = F.Material(
>     D_0={protium: 1.0e-3, deuterium: 1.0e-3, tritium: 1.0e-3},  # example diffusivity values for H, D, T
>     E_D={protium: 0.0, deuterium: 0.0, tritium: 0.0},
> )
> 
> # One bulk domain using this material
> bulk = F.VolumeSubdomain1D(id=1, borders=[0, 1], material=mat)
> 
> my_model.subdomains = [bulk, left_surf, right_surf]
> # Boundary conditions
> my_model.boundary_conditions = [
>     F.FixedConcentrationBC(left_surf, value=1, species=protium),
>     F.FixedConcentrationBC(right_surf, value=0, species=protium),
>     F.FixedConcentrationBC(left_surf, value=1, species=deuterium),
>     F.FixedConcentrationBC(right_surf, value=0, species=deuterium),
>     F.FixedConcentrationBC(left_surf, value=1, species=tritium),
>     F.FixedConcentrationBC(right_surf, value=0, species=tritium),
> ]
> 
> my_model.temperature = 300
> my_model.settings = F.Settings(atol=1e-10, rtol=1e-10, final_time=5)
> my_model.settings.stepsize = F.Stepsize(1)

Author: No need to redefine all of this here, simply overwrite `mat.D_0` and re-initialise the model. Also no need to hide this cell
```suggestion
mat.D_0 = {protium: 1.0e-3, deuterium: 1.0e-3, tritium: 1.0e-3}
```

my_model.initialise()
my_model.run()

> import matplotlib.pyplot as plt

Author: Same comment as above, split this code cell here and only hide the matplotlib stuff

def plot_profile(species, **kwargs):
    c = species.post_processing_solution.x.array[:]
    x = species.post_processing_solution.function_space.mesh.geometry.x[:,0]
    return plt.plot(x, c, **kwargs)

for species in my_model.species:
    plot_profile(species, label=species.name)

plt.xlabel('Position')
plt.ylabel('Concentration')
plt.legend()
plt.show()
```

> ## Solving a problem with custom material properties ##

Author: ```suggestion
## Custom diffusion coefficient
```

In this 2D example, we show how to define a spatially-dependent diffusivity. Specifcally, we'll look at a circular diffusivity profile: $ D = 0.1 + x^2 + y^2 $

First, we need to define a 2D mesh:

```{code-cell} ipython3
import dolfinx
from dolfinx.mesh import create_unit_square
from mpi4py import MPI
import numpy as np
import festim as F
from basix.ufl import element
from dolfinx import plot
import pyvista

mesh_fenics = create_unit_square(MPI.COMM_WORLD, 20, 20)
my_model = F.HydrogenTransportProblem()
my_model.mesh = F.Mesh(mesh_fenics)
```

Users can define a spatially-dependent material property using `fem.Function`, which requires a function space (with the correct order) to be setup using `fem.functionspace`:

```{code-cell} ipython3
el = element("Lagrange", mesh_fenics.topology.cell_name(), 2)
V = dolfinx.fem.functionspace(my_model.mesh.mesh, el)

# Define diffusion coefficient as scalar field
diffusivity = dolfinx.fem.Function(V)
diffusivity.interpolate(lambda x: 0.1 + x[0]**2 + x[1]**2)
```

> We can visualize the diffusivity and contours (to verify the circular profile):

Author: ```suggestion
Here's our diffusivity field:
```

> ```{code-cell} ipython3
> topology, cell_types, geometry = plot.vtk_mesh(V)
> function_grid = pyvista.UnstructuredGrid(topology, cell_types, geometry)

Author: ```suggestion
```{code-cell} ipython3
:tags: [hide-input]

pyvista.start_xvfb()
pyvista.set_jupyter_backend("html")

topology, cell_types, geometry = plot.vtk_mesh(V)
function_grid = pyvista.UnstructuredGrid(topology, cell_types, geometry)
```

function_grid.point_data["D"] = diffusivity.x.array.real
function_grid.set_active_scalars("D")

# Generate contours (e.g. 10 levels)
contours = function_grid.contour(isosurfaces=5, scalars="D")

plotter = pyvista.Plotter()
plotter.add_mesh(function_grid, cmap="viridis", show_edges=False, opacity=1)
plotter.add_mesh(contours, color="r", line_width=0.5)  # contour lines
plotter.view_xy()
plotter.add_text("Spatially varying diffusivity", font_size=12)

if not pyvista.OFF_SCREEN:
    plotter.show()
```

> Now we can add this property directly to a FESTIM `Material`:

Author: We should add a [note ](https://jupyterbook.org/en/stable/content/content-blocks.html#notes-warnings-and-other-admonitions) here to point out that it's passed to the `D` argument and not `D_0`
